File: frontend/lib/learningXpEngine.js
/**
 * Learning Hub XP Engine
 * 
 * Pure utility functions for XP calculation, streak logic,
 * anti-farming protection, and unlock progression.
 * 
 * All progression is SUBJECT-SPECIFIC — no global inheritance.
 */

import { format, differenceInCalendarDays } from 'date-fns';

// ============================
// Constants
// ============================
export const BASE_XP_PER_TOPIC = 7;
export const MAX_XP_ELIGIBLE_TOPICS_PER_MODULE = 10;
export const MODULE_COMPLETION_MULTIPLIER = 5;
export const CHAPTER_COMPLETION_MULTIPLIER = 10;

const STREAK_TIERS = [
  { days: 90, bonus: 7 },
  { days: 21, bonus: 3 },
  { days: 7,  bonus: 1 },
];

// ============================
// Streak Calculations
// ============================

/**
 * Get the streak bonus XP based on current subject streak count.
 * @param {number} streakCount - current subject streak days
 * @returns {number} bonus XP per topic
 */
export function getStreakBonus(streakCount) {
  for (const tier of STREAK_TIERS) {
    if (streakCount >= tier.days) return tier.bonus;
  }
  return 0;
}

/**
 * Get the effective XP per topic for a subject's current streak.
 * @param {number} streakCount
 * @returns {number}
 */
export function getXpPerTopic(streakCount) {
  return BASE_XP_PER_TOPIC + getStreakBonus(streakCount);
}

/**
 * Get the current streak tier label.
 * @param {number} streakCount
 * @returns {{ label: string, color: string, nextTier: number|null }}
 */
export function getStreakTier(streakCount) {
  if (streakCount >= 90) return { label: 'Legendary', color: '#FF6B2C', nextTier: null };
  if (streakCount >= 21) return { label: 'Dedicated', color: '#FFD700', nextTier: 90 };
  if (streakCount >= 7)  return { label: 'Committed', color: '#3B82F6', nextTier: 21 };
  if (streakCount >= 1)  return { label: 'Active', color: '#10B981', nextTier: 7 };
  return { label: 'New', color: '#94A3B8', nextTier: 1 };
}

/**
 * Calculate updated streak for a subject when a topic is completed today.
 * @param {number} currentStreak - existing streak
 * @param {string|null} lastStreakDate - ISO date string of last streak activity
 * @returns {{ newStreak: number, isStreakMilestone: number|null }}
 */
export function calculateUpdatedStreak(currentStreak, lastStreakDate) {
  const today = format(new Date(), 'yyyy-MM-dd');
  
  // Already maintained today
  if (lastStreakDate === today) {
    return { newStreak: currentStreak, isStreakMilestone: null };
  }

  let newStreak;
  
  if (!lastStreakDate) {
    // First ever activity
    newStreak = 1;
  } else {
    const daysDiff = differenceInCalendarDays(new Date(today), new Date(lastStreakDate));
    if (daysDiff === 1) {
      // Consecutive day
      newStreak = currentStreak + 1;
    } else if (daysDiff === 0) {
      // Same day (shouldn't reach here but safety)
      newStreak = currentStreak;
    } else {
      // Streak broken — restart at 1
      newStreak = 1;
    }
  }

  // Check if this hits a milestone
  const milestones = [7, 21, 90];
  const isStreakMilestone = milestones.includes(newStreak) ? newStreak : null;

  return { newStreak, isStreakMilestone };
}

// ============================
// Anti-Farming Protection
// ============================

/**
 * Check if a topic completion should award XP.
 * @param {string} topicId - ID of the topic being completed
 * @param {string} moduleId - ID of the parent module (chapter)
 * @param {Object} completedTopicsLog - { [moduleId]: number } map of XP-awarded topic counts per module
 * @param {Object} xpAwardedTopics - { [topicId]: true } map of already-awarded topics
 * @returns {{ eligible: boolean, reason: string }}
 */
export function checkXpEligibility(topicId, moduleId, completedTopicsLog, xpAwardedTopics) {
  // Already awarded XP for this specific topic
  if (xpAwardedTopics[topicId]) {
    return { eligible: false, reason: 'XP already awarded for this topic' };
  }

  // Check per-module cap
  const moduleCount = completedTopicsLog[moduleId] || 0;
  if (moduleCount >= MAX_XP_ELIGIBLE_TOPICS_PER_MODULE) {
    return { eligible: false, reason: `Module XP cap reached (${MAX_XP_ELIGIBLE_TOPICS_PER_MODULE} topics)` };
  }

  return { eligible: true, reason: 'Eligible' };
}

// ============================
// XP Award Calculations
// ============================

/**
 * Calculate XP to award for a topic completion.
 * @param {number} streakCount - current subject streak
 * @param {string} topicId
 * @param {string} moduleId
 * @param {Object} completedTopicsLog
 * @param {Object} xpAwardedTopics
 * @returns {{ xp: number, eligible: boolean, reason: string, xpPerTopic: number }}
 */
export function calculateTopicXp(streakCount, topicId, moduleId, completedTopicsLog, xpAwardedTopics) {
  const xpPerTopic = getXpPerTopic(streakCount);
  const { eligible, reason } = checkXpEligibility(topicId, moduleId, completedTopicsLog, xpAwardedTopics);
  
  return {
    xp: eligible ? xpPerTopic : 0,
    eligible,
    reason,
    xpPerTopic
  };
}

/**
 * Calculate module completion bonus.
 * @param {number} streakCount
 * @returns {number}
 */
export function calculateModuleBonus(streakCount) {
  return getXpPerTopic(streakCount) * MODULE_COMPLETION_MULTIPLIER;
}

/**
 * Calculate chapter completion bonus.
 * @param {number} streakCount
 * @returns {number}
 */
export function calculateChapterBonus(streakCount) {
  return getXpPerTopic(streakCount) * CHAPTER_COMPLETION_MULTIPLIER;
}

// ============================
// Completion Checks
// ============================

/**
 * Check if a module (chapter) is fully completed.
 * @param {Array} topics - array of topic objects
 * @returns {boolean}
 */
export function isModuleComplete(topics) {
  if (!topics || topics.length === 0) return false;
  return topics.every(t => t.status === 'Completed');
}

/**
 * Check if a chapter (collection of modules) is fully completed.
 * All child modules must be complete.
 * @param {Array} chapters - array of chapter objects with topics
 * @returns {boolean}
 */
export function isChapterComplete(chapters) {
  if (!chapters || chapters.length === 0) return false;
  return chapters.every(ch => isModuleComplete(ch.topics));
}

/**
 * Check if the previous module in sequence is complete for soft unlock.
 * @param {Array} chapters - ordered array of chapters
 * @param {number} targetIndex - index of the module being opened
 * @returns {{ locked: boolean, previousTitle: string|null }}
 */
export function checkModuleSoftLock(chapters, targetIndex) {
  if (targetIndex === 0) return { locked: false, previousTitle: null };
  
  const previousChapter = chapters[targetIndex - 1];
  if (!previousChapter) return { locked: false, previousTitle: null };
  
  const isComplete = isModuleComplete(previousChapter.topics);
  return {
    locked: !isComplete,
    previousTitle: previousChapter.title
  };
}

// ============================
// Subject Stats Calculator
// ============================

/**
 * Compute aggregate stats for the Learning Hub.
 * @param {Array} subjects - all subjects from store
 * @returns {Object}
 */
export function computeLearningStats(subjects) {
  let totalXp = 0;
  let totalCompletedTopics = 0;
  let totalTopics = 0;
  let totalCompletedModules = 0;
  let totalModules = 0;
  let maxStreak = 0;
  let totalStudyHours = 0;

  for (const subject of subjects) {
    totalXp += subject.xpEarned || 0;
    totalStudyHours += subject.totalHours || 0;
    if ((subject.streakCount || 0) > maxStreak) maxStreak = subject.streakCount || 0;

    for (const chapter of (subject.chapters || [])) {
      totalModules++;
      const topics = chapter.topics || [];
      totalTopics += topics.length;
      const completed = topics.filter(t => t.status === 'Completed').length;
      totalCompletedTopics += completed;
      if (topics.length > 0 && completed === topics.length) totalCompletedModules++;
    }
  }

  return {
    totalXp,
    totalCompletedTopics,
    totalTopics,
    totalCompletedModules,
    totalModules,
    maxStreak,
    totalStudyHours
  };
}

export default {
  getStreakBonus,
  getXpPerTopic,
  getStreakTier,
  calculateUpdatedStreak,
  checkXpEligibility,
  calculateTopicXp,
  calculateModuleBonus,
  calculateChapterBonus,
  isModuleComplete,
  isChapterComplete,
  checkModuleSoftLock,
  computeLearningStats,
  BASE_XP_PER_TOPIC,
  MAX_XP_ELIGIBLE_TOPICS_PER_MODULE,
};

File: mobile/lib/core/learning/learning_xp_engine.dart
import 'package:flutter/material.dart';
import '../../models/learning_model.dart';

class StreakTier {
  final String label;
  final Color color;
  final int? nextTier;

  const StreakTier({
    required this.label,
    required this.color,
    this.nextTier,
  });
}

class StreakUpdateResult {
  final int newStreak;
  final int? isStreakMilestone;

  StreakUpdateResult({
    required this.newStreak,
    this.isStreakMilestone,
  });
}

class XpEligibilityResult {
  final bool eligible;
  final String reason;

  XpEligibilityResult({
    required this.eligible,
    required this.reason,
  });
}

class TopicXpResult {
  final int xp;
  final bool eligible;
  final String reason;
  final int xpPerTopic;

  TopicXpResult({
    required this.xp,
    required this.eligible,
    required this.reason,
    required this.xpPerTopic,
  });
}

class SoftLockResult {
  final bool locked;
  final String? previousTitle;

  SoftLockResult({
    required this.locked,
    this.previousTitle,
  });
}

class LearningStats {
  final int totalXp;
  final int totalCompletedTopics;
  final int totalTopics;
  final int totalCompletedModules;
  final int totalModules;
  final int maxStreak;
  final int totalStudyHours;

  LearningStats({
    required this.totalXp,
    required this.totalCompletedTopics,
    required this.totalTopics,
    required this.totalCompletedModules,
    required this.totalModules,
    required this.maxStreak,
    required this.totalStudyHours,
  });
}

class LearningXpEngine {
  static const int baseXpPerTopic = 7;
  static const int maxXpEligibleTopicsPerModule = 10;
  static const int moduleCompletionMultiplier = 5;
  static const int chapterCompletionMultiplier = 10;

  static const List<Map<String, dynamic>> streakTiersConfig = [
    {'days': 90, 'bonus': 7},
    {'days': 21, 'bonus': 3},
    {'days': 7,  'bonus': 1},
  ];

  static int getStreakBonus(int streakCount) {
    for (final tier in streakTiersConfig) {
      if (streakCount >= (tier['days'] as int)) {
        return tier['bonus'] as int;
      }
    }
    return 0;
  }

  static int getXpPerTopic(int streakCount) {
    return baseXpPerTopic + getStreakBonus(streakCount);
  }

  static StreakTier getStreakTier(int streakCount) {
    if (streakCount >= 90) {
      return const StreakTier(label: 'Legendary', color: Color(0xFFE25B20), nextTier: null);
    }
    if (streakCount >= 21) {
      return const StreakTier(label: 'Dedicated', color: Color(0xFFFFD700), nextTier: 90);
    }
    if (streakCount >= 7) {
      return const StreakTier(label: 'Committed', color: Color(0xFF3B82F6), nextTier: 21);
    }
    if (streakCount >= 1) {
      return const StreakTier(label: 'Active', color: Color(0xFF10B981), nextTier: 7);
    }
    return const StreakTier(label: 'New', color: Color(0xFF94A3B8), nextTier: 1);
  }

  static String formatDate(DateTime dt) {
    return "${dt.year.toString().padLeft(4, '0')}-${dt.month.toString().padLeft(2, '0')}-${dt.day.toString().padLeft(2, '0')}";
  }

  static int differenceInCalendarDays(DateTime a, DateTime b) {
    final aLocal = DateTime(a.year, a.month, a.day);
    final bLocal = DateTime(b.year, b.month, b.day);
    return aLocal.difference(bLocal).inDays;
  }

  static StreakUpdateResult calculateUpdatedStreak(int currentStreak, String? lastStreakDateStr) {
    final todayStr = formatDate(DateTime.now());
    
    if (lastStreakDateStr == todayStr) {
      return StreakUpdateResult(newStreak: currentStreak, isStreakMilestone: null);
    }

    int newStreak;

    if (lastStreakDateStr == null || lastStreakDateStr.isEmpty) {
      newStreak = 1;
    } else {
      try {
        final lastDate = DateTime.parse(lastStreakDateStr);
        final today = DateTime.parse(todayStr);
        final daysDiff = differenceInCalendarDays(today, lastDate);

        if (daysDiff == 1) {
          newStreak = currentStreak + 1;
        } else if (daysDiff == 0) {
          newStreak = currentStreak;
        } else {
          newStreak = 1;
        }
      } catch (e) {
        newStreak = 1;
      }
    }

    final milestones = [7, 21, 90];
    final isStreakMilestone = milestones.contains(newStreak) ? newStreak : null;

    return StreakUpdateResult(newStreak: newStreak, isStreakMilestone: isStreakMilestone);
  }

  static XpEligibilityResult checkXpEligibility(
    String topicId,
    String moduleId,
    Map<String, int> completedTopicsLog,
    Map<String, bool> xpAwardedTopics,
  ) {
    if (xpAwardedTopics[topicId] == true) {
      return XpEligibilityResult(eligible: false, reason: 'XP already awarded for this topic');
    }

    final moduleCount = completedTopicsLog[moduleId] ?? 0;
    if (moduleCount >= maxXpEligibleTopicsPerModule) {
      return XpEligibilityResult(
        eligible: false, 
        reason: 'Module XP cap reached ($maxXpEligibleTopicsPerModule topics)',
      );
    }

    return XpEligibilityResult(eligible: true, reason: 'Eligible');
  }

  static TopicXpResult calculateTopicXp(
    int streakCount,
    String topicId,
    String moduleId,
    Map<String, int> completedTopicsLog,
    Map<String, bool> xpAwardedTopics,
  ) {
    final xpPerTopic = getXpPerTopic(streakCount);
    final eligibility = checkXpEligibility(topicId, moduleId, completedTopicsLog, xpAwardedTopics);

    return TopicXpResult(
      xp: eligibility.eligible ? xpPerTopic : 0,
      eligible: eligibility.eligible,
      reason: eligibility.reason,
      xpPerTopic: xpPerTopic,
    );
  }

  static int calculateModuleBonus(int streakCount) {
    return getXpPerTopic(streakCount) * moduleCompletionMultiplier;
  }

  static int calculateChapterBonus(int streakCount) {
    return getXpPerTopic(streakCount) * chapterCompletionMultiplier;
  }

  static bool isModuleComplete(List<TopicModel>? topics) {
    if (topics == null || topics.isEmpty) return false;
    return topics.every((t) => t.status == 'Completed');
  }

  static bool isChapterComplete(List<ChapterModel>? chapters) {
    if (chapters == null || chapters.isEmpty) return false;
    return chapters.every((ch) => isModuleComplete(ch.topics));
  }

  static SoftLockResult checkModuleSoftLock(List<ChapterModel> chapters, int targetIndex) {
    if (targetIndex <= 0) return SoftLockResult(locked: false);
    
    final previousChapter = chapters[targetIndex - 1];
    final isComplete = isModuleComplete(previousChapter.topics);
    
    return SoftLockResult(
      locked: !isComplete,
      previousTitle: previousChapter.title,
    );
  }

  static LearningStats computeLearningStats(List<SubjectModel> subjects) {
    int totalXp = 0;
    int totalCompletedTopics = 0;
    int totalTopics = 0;
    int totalCompletedModules = 0;
    int totalModules = 0;
    int maxStreak = 0;
    int totalStudyHours = 0;

    for (final subject in subjects) {
      totalXp += subject.xpEarned;
      totalStudyHours += subject.totalHours ?? 0;
      if (subject.streakCount > maxStreak) {
        maxStreak = subject.streakCount;
      }

      final chapters = subject.chapters ?? [];
      for (final chapter in chapters) {
        totalModules++;
        final topics = chapter.topics ?? [];
        totalTopics += topics.length;
        
        final completed = topics.where((t) => t.status == 'Completed').length;
        totalCompletedTopics += completed;

        if (topics.isNotEmpty && completed == topics.length) {
          totalCompletedModules++;
        }
      }
    }

    return LearningStats(
      totalXp: totalXp,
      totalCompletedTopics: totalCompletedTopics,
      totalTopics: totalTopics,
      totalCompletedModules: totalCompletedModules,
      totalModules: totalModules,
      maxStreak: maxStreak,
      totalStudyHours: totalStudyHours,
    );
  }
}
